torch/distributed/_functional_collectives.py

```diff
@@ -0,0 +1,231 @@
+first usage of the tensor and insert cross stream sync at the right place.
+
+The above are the easy bit, the hard ones are how we match the Work object returned by
+c10d and the tensor AsyncCollectiveTensor wraps. We alloc the tensor inside the collective
+op implementation (see ``clone()`` call in ``_all_reduce``) and then it goes into a voyage
+of wonders and magic through the dispatcher and eventually something comes out of it.
+
+The tensor the caller of an op receives is not guaranteed to be the same allocated by an
+op implenentation as mechanisms like vmap or autograd can produce different instances -
+even if they ultimately point to the same underlying data.
+
+Given those constraints we employ the following schema.
+
+We have a dictionary of tensor::data_ptr -> Work that we insert right after we call into c10d.
+
+We use this dictionary during the witness callback to invoke Work::wait()
+
+Finally, we setup a finalizer against the tensor wrapper to observe it getting collected so we
+can clean up stale entries in the dictionary.
+
+To eliminate the possiblity of races we have a global version counter that is used by the finalizer.
+
+As a wise man said once: Don't cross the streams (https://www.youtube.com/watch?v=wyKQe_i9yyo)
+
+"""
+data_ptr_to_work = dict()
+work_version = 0
+
+def _register_tensor_inner(tensor, work):
+    global data_ptr_to_work
+    global work_version
+    data_ptr_to_work[tensor.data_ptr()] = (work_version, work)
+    work_version += 1
+
+def _clear_tensor(data_ptr, version):
+    global data_ptr_to_work
+    version_and_work = data_ptr_to_work.get(data_ptr)
+
+    if version_and_work is not None and version_and_work[0] == version:
+        del data_ptr_to_work[data_ptr]
+
+def _register_tensor_wrapper_inner(tensor_wrapper, tensor):
+    global data_ptr_to_work
+    version, _ = data_ptr_to_work.get(tensor.data_ptr(), (None, None))
+    if version is None:
+        warnings.warn("Trying to register finalizers to AsyncCollectiveTensor but the inner tensor is already gone")
+    else:
+        weakref.finalize(tensor_wrapper, _clear_tensor, tensor.data_ptr(), version)
+
+def _wait_tensor_inner(tensor):
+    global data_ptr_to_work
+    version_and_work = data_ptr_to_work.get(tensor.data_ptr())
+    if version_and_work is not None:
+        version_and_work[1].wait()
+        del data_ptr_to_work[tensor.data_ptr()]
+
+lib = torch.library.Library("tr_c10d", "DEF")
+lib.define("wait(Tensor self) -> Tensor")
+
+impl_lib_cpu = torch.library.Library("tr_c10d", "IMPL", "CPU")
+impl_lib_cuda = torch.library.Library("tr_c10d", "IMPL", "CUDA")
+
+def _wait_tensor(tensor: torch.Tensor):
+    _wait_tensor_inner(tensor)
+    return tensor
+
+impl_lib_cpu.impl("wait", _wait_tensor)
+impl_lib_cuda.impl("wait", _wait_tensor)
+
+def wait_tensor(tensor):
+    return torch._ops.ops.tr_c10d.wait(tensor)
+
+class AsyncCollectiveTensor(torch.Tensor):
```

Comment:
Is it worse to call this just AsyncTensor?  I realized the name is long and it's a mouthful, but the name is also precise and correct so it might be good to leave it